### ISSUE TITLE
Boolean pattern matches

### ISSUE THREAD

**Issue by nex3** [dart-lang/language#3062]
As part of converting Dart Sass to Dart 3 style, I've noticed several cases where I wish I could use pattern matching _just_ to express boolean conditions without any additional variable bindings. For example, I'd _like_ to be able to write something like

```dart
node matches ListExpression(
  separator: ListSeparator.comma,
  hasBrackets: false,
  contents: [_, _, ...]
)
```

instead of

```dart
node is ListExpression &&
    node.separator == ListSeparator.comma &&
    !node.hasBrackets &&
    node.contents.length > 1
```

...but there's not a great way to do that. The best alternative is something much more awkward like

```dart
switch (node) {
    ListExpression(
        separator: ListSeparator.comma,
        hasBrackets: false,
        contents: [_, _, ...]
    ) => true,
    _ => false
}
```

**Comment by munificent** [comment]
Yes, this is a cool idea. The obvious syntax would probably be `<expr> case <pattern>`, like the head of an `if-case` statement:

```dart
var isNonEmptyList = this.node case ListExpression(
        separator: ListSeparator.comma,
        hasBrackets: false,
        contents: [_, _, ...]
    );
```

The hard part is figuring out the precedence and determining when the pattern on the right is done and any surrounding expression begins. The expression and pattern grammars are both quite complex, so the current design works pretty hard to keep them clearly partitioned from each other so that the parser (and reader) can always tell when one ends and the next begins.

If we limited the pattern after `case` to be an explicitly delimited one (record, list, object, map, or parentheses), that would probably do it.

**Comment by lrhn** [comment]
Binding variables is fine, but the `when` clause would be the only place they are in scope.
Except if you use it in a condition where code is guarded by the test. This would/should allow:
```dart
if ((e1 case int x) && (e2 case int y)) { return x + y; }
```


Promotion due to pattern matching would still work through the boolean variablen, like we do for other type checks today.

I'm not too worried about delimiting the syntax, if we can just require parentheses in most contexts where there could be any doubt.

**Comment by nex3** [comment]
As an additional note: if this is added, it would be nice to also have `!case` or similar syntax for the negation, to avoid the awkward `!(name case ... )` form.

**Comment by HENRDS** [comment]
I don't know about the feasibility of it in Dart, or if it has already been proposed, but a syntax similar to C#'s could be an option.

It would basically consist of allowing patterns after `is` and `is!`:
```dart
if (this.node is ListExpression(separator: ListSeparator.comma, hasBrackets: false, contents: [_, _, ...]))
{
    // ...
}


if (name is! String) {
    // ...
}
```

Binding the result to a new variable could simulate `when`:
```dart
if (x is Foo(bar: Bar.optionA) foo && (foo.baz == 1 || foo.baz == 2)) {
    // ...
}
```

**Comment by HENRDS** [comment]
@munificent, that makes sense, thanks for explaining.
